rustls/src/conn.rs

```diff
@@ -719,13 +700,177 @@ impl ConnectionCommon {
+            self.common_state
+                .send_fatal_alert(AlertDescription::IllegalParameter);
+        }
+
+        // If we get a CloseNotify, make a note to declare EOF to our
+        // caller.
+        if alert.description == AlertDescription::CloseNotify {
+            self.peer_eof = true;
+            return Ok(());
+        }
+
+        // Warnings are nonfatal for TLS1.2, but outlawed in TLS1.3
+        // (except, for no good reason, user_cancelled).
+        if alert.level == AlertLevel::Warning {
+            if self.common_state.is_tls13() && alert.description != AlertDescription::UserCanceled {
+                self.common_state
+                    .send_fatal_alert(AlertDescription::DecodeError);
+            } else {
+                warn!("TLS alert warning received: {:#?}", alert);
+                return Ok(());
+            }
+        }
+
+        error!("TLS alert received: {:#?}", alert);
+        Err(Error::AlertReceived(alert.description))
+    }
+
+
+    /// Are we done? i.e., have we processed all received messages,
+    /// and received a close_notify to indicate that no new messages
+    /// will arrive?
+    pub fn connection_at_eof(&self) -> bool {
+        self.peer_eof && !self.message_deframer.has_pending()
+    }
+
+    /// Read TLS content from `rd`.  This method does internal
+    /// buffering, so `rd` can supply TLS messages in arbitrary-
+    /// sized chunks (like a socket or pipe might).
+    pub fn read_tls(&mut self, rd: &mut dyn io::Read) -> io::Result<usize> {
+        self.message_deframer.read(rd)
+    }
+
+    /// Send plaintext application data, fragmenting and
+    /// encrypting it as it goes out.
+    ///
+    /// If internal buffers are too small, this function will not accept
+    /// all the data.
+    pub fn send_some_plaintext(&mut self, data: &[u8]) -> usize {
+        self.common_state
+            .send_plain(data, Limit::Yes)
+    }
+
+    pub fn send_early_plaintext(&mut self, data: &[u8]) -> usize {
+        debug_assert!(self.common_state.early_traffic);
+        debug_assert!(
+            self.common_state
+                .record_layer
+                .is_encrypting()
+        );
+
+        if data.is_empty() {
+            // Don't send empty fragments.
+            return 0;
+        }
+
+        self.common_state
+            .send_appdata_encrypt(data, Limit::Yes)
+    }
+
+    pub fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let len = self
+            .common_state
+            .received_plaintext
+            .read(buf)?;
+
+        if len == 0 && !buf.is_empty() {
+            // no bytes available:
+            // - if we received a close_notify, this is a genuine permanent EOF
+            // - otherwise say EWOULDBLOCK
+            if !self.connection_at_eof() {
+                return Err(io::ErrorKind::WouldBlock.into());
+            }
+        }
+
+        Ok(len)
+    }
+}
+
+pub(crate) struct CommonState {
```

Comment:
I don't love this name but couldn't think of anything better. This is the connection state shared by client and server, but is different from the state machine types that implement the `State` trait.